applications/shell_2/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+                Ok(())
+            }
+            Err(Error::CommandNotFound(command)) => {
+                println!("{}: command not found", command);
+                Ok(())
+            }
+            _ => Ok(()),
+        }
+    }
+
+    // TODO: Use guards?
+    fn execute_external(&mut self, cmd: &str, args: Vec<&str>) -> Result<()> {
+        self.set_app_discipline();
+        let result = self._execute_external(cmd, args);
+        self.set_shell_discipline();
+        result
+    }
+
+    fn _execute_external(&mut self, cmd: &str, args: Vec<&str>) -> Result<()> {
+        let mut job = self.resolve_external(cmd, args)?;
+
+        let mut num = 1;
+        while self.jobs.contains_key(&num) {
+            num += 1;
+        }
+
+        job.unsuspend();
+        let job = self.jobs.try_insert(num, job).unwrap();
+        self.discipline.clear_events();
+
+        loop {
+            if let Ok(event) = self.discipline.event_receiver().try_receive() {
+                return match event {
+                    Event::CtrlC => {
+                        job.kill();
+                        self.jobs.remove(&num).unwrap();
+                        Err(Error::Command(130))
+                    }
+                    Event::CtrlD => todo!(),
+                    Event::CtrlZ => {
+                        job.suspend();
+                        todo!();
+                    }
+                };
+            } else if let Some(exit_value) = job.update() {
+                self.jobs.remove(&num).unwrap();
+                return match exit_value {
+                    0 => Ok(()),
+                    _ => Err(Error::Command(exit_value)),
+                };
+            }
+        }
+    }
+
+    fn resolve_external(&self, cmd: &str, args: Vec<&str>) -> Result<Job> {
+        let namespace_dir = task::get_my_current_task()
+            .map(|t| t.get_namespace().dir().clone())
+            .expect("couldn't get namespace dir");
+
+        let crate_name = format!("{}-", cmd);
+        let mut matching_files = namespace_dir
+            .get_files_starting_with(&crate_name)
+            .into_iter();
+
+        let app_path = match matching_files.next() {
+            Some(f) => Path::new(f.lock().get_absolute_path()),
+            None => return Err(Error::CommandNotFound(cmd.to_owned())),
+        };
+
+        if matching_files.next().is_some() {
+            panic!("multiple matching files found");
+        }
+
+        let task = spawn::new_application_task_builder(app_path, None)
+            .unwrap()
+            .argument(
+                args.into_iter()
+                    .map(|arg| arg.to_owned())
+                    .collect::<Vec<_>>(),
+            )
+            .block()
+            .spawn()
+            .unwrap();
+
+        let id = task.id;
+        // TODO: Double arc :(
+        app_io::insert_child_streams(id, app_io::streams().unwrap());
+        // TODO: set environment
```

Comment:
I think this can be delayed till later. The double `Arc` is unfortunate but not too important. I haven't set the environment because currently, it behaves in a weird way where the child can modify the parent's environment. I think this should be changed but left it for another PR.

Reply:
right, the environment is shared with that of the "parent" task. If you don't want that behavior, you can clone the environment, set the child task's env to that new cloned instance, and then modify it.

Reply:
Turns out this is already done in `Task::new`. The old shell was intentionally overwriting the child's environment with a reference to the parent's environment.